### PR TITLE
gh-90155: Fix broken `asyncio.Semaphore` and strengthen FIFO guarantee.

### PR DESCRIPTION
**Pull request by cykerway**
# gh-90155: Fix broken :class:`asyncio.Semaphore` and strengthen FIFO guarantee.

Current `asyncio.Semaphore` may become broken on certain workflow. Tasks waiting on a broken `Semaphore` can hang forever. This PR not only fixes this problem but also strengthens the FIFO guarantee on `Semaphore` waiters. Test cases show details.

### CODE REVIEW

**Comment by mguentner** [comment]
@cykerway Thank you for your work on this. I can confirm that this is indeed an issue in the wild.

I spend quite some time to debug an application that was blocking in rare cases after seeing cancelled tasks. Finally I checked the internal state of the `Semaphore` and saw that `_wakeup_scheduled` was `True` with no `_waiters` left to set it to `False` which blocks it forever.

Before I found this PR, I came up with another solution which I have documented [here](https://github.com/mguentner/async_semaphore_race_condition_poc) together with a program that reproduces the race condition (inspired by the tests added by this PR).

File: Lib/asyncio/locks.py
Comment (gvanrossum) [inline Lib/asyncio/locks.py]: Let's always return a `bool`, like before:
```suggestion
        return self._value <= 0 or bool(self._wakeup_scheduled)
```

File: Lib/asyncio/locks.py
Comment (gvanrossum) [inline Lib/asyncio/locks.py]: People outside the US don't know what "freshman" means (I know I didn't). Also it feels ever so slightly pejorative.
```suggestion
        first = True
        while self._value <= 0 or (first and self._wakeup_scheduled):
            first = False
```

**Review by gvanrossum — COMMENTED** [review]
I'm not going to lie, I don't understand this code well enough to approve it. :-(

I wonder where we went wrong in asyncio's design that the simple version (which has long been fixed and fixed again) didn't work. :-(

**Comment by cykerway** [comment]
> I wonder where we went wrong in asyncio's design that the simple version (which has long been fixed and fixed again) didn't work. :-(

(If I remembered correctly)

1. The original implementation before gh-90155 was working but just not fifo. But the [doc](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) doesn't say semaphores are fair or fifo. People who don't need fifo semaphores can use the original one.

2. Someone was unsatisfied about task starvation and opened gh-90155. Changes introduced by it (such as https://github.com/python/cpython/commit/9d59381a5d20157930bae34e5f5a7bc5ef09fa89) were aimed to bring fairness to semaphores, but the implementation was flawed and introduced regression that makes the semaphore unusable in certain race conditions. This is what I showed in https://github.com/python/cpython/issues/90155#issuecomment-1137052110

3. Those patches have been officially merged into 3.10 and other versions. So semaphores are _currently broken_. I mean the python installed by system package managers, not nightly build here. This includes current version 3.10.7.

4. This PR is meant to be a hotfix to this problem. It doesn't mean to address fifo or non-fifo problem but tries to bring back a usable semaphore with minimal change on the existing implementation. Another way is to just revert everything to the very beginning, but I think the advantage of taking this PR than reverting everything is to prevent task starvation. And in either case the tests added by this PR are still useful in preventing future regressions. You are welcome to add more tests to ensure this PR itself doesn't bring more regressions.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: If this is supposed to reproduce the example reproducer from the issue, I'd expect c1 to sleep a tiny fraction (0.01 or so) and c2 a long time (say 10 sec) since if the fix works, c2 will be interrupte by c1 anyways. Sleeping zero doesn't obviously guarantee that c2 gets to run until it hits the await, nor that c2 is still around when c1 cancels it.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: Can we just use a nonlocal variable instead of a list of tasks being passed around? c2 doesn't need to know about c1, c1 just needs to know about the c2 task.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: Would be nice to assert the return value -- what *do* you expect this to return?

File: Misc/NEWS.d/next/Library/2022-05-25-15-57-39.gh-issue-90155.YMstB5.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2022-05-25-15-57-39.gh-issue-90155.YMstB5.rst]: Can we qualify that it's broken when an acquire() operation is cancelled? And I'm not sure that we're actually strengthening the FIFO guarantee -- apart from the cancellation bug it was perfectly FIFO already, right?

**Review by gvanrossum — COMMENTED** [review]
I approve of the new solution. I have some nits about the tests, not too sure about those so push back if you think I misunderstand how it works!

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: Isn't that c2 doesn't hit the await?

c1 acquires the sem first, then gives control to c2 with `sleep(0)`. c2 gets the control but not the sem, so c2 is blocked at `async with sem`, which gives control to c1. c1 then releases the sem. Release is a normal function, and c1 doesn't yield before cancel c2. So c2 is still around when c1 cancels it?

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: This is doable.

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: `[ None, <CancelledError> ]`

Exceptions don't compare equal so just check its type.

File: Misc/NEWS.d/next/Library/2022-05-25-15-57-39.gh-issue-90155.YMstB5.rst
Comment (cykerway) [reply Misc/NEWS.d/next/Library/2022-05-25-15-57-39.gh-issue-90155.YMstB5.rst]: The `Lock` docs say `Lock` is FIFO. I didn't come up with a proof that the adapted `Semaphore` is FIFO (although I believe so). Yet the `Semephore` docs do not say it is FIFO, so we can't violate its docs.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_locks.py]: Okay, that's solid, as long as event loops guarantee that tasks are started in the order created. I'm not sure that we _guarantee_ that, and although I expect that if we changed the behavior we'd get lots of failed tests, if we want to ensure that c1 gets the semaphore, the main test function should probably `sleep(0)` between the task creations. And it sounds like the sleep in c2 is never reached, so might as well put `self.assertFalse()` there.

File: Lib/asyncio/locks.py
Comment (cykerway) [inline Lib/asyncio/locks.py]: What do you think about this statement? This method obviously doesn't take into account waiters.

The `Lock` class states it as

> Return True if the lock is locked.

Not being locked doesn't necessarily mean it can be acquired immediately.

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: Not sure if you have seen the questions I posted in the issue https://github.com/python/cpython/issues/90155#issuecomment-1250136232

I think it helps if we have clear answers about the event loop model. In numerous places in the tests I felt it was assumed that tasks are started in the order created. Yet there is no confirmation of that. Also, I was forced to tinker with sleep time. The scheduling always looks a puzzle to me.

`self.assertFalse()` is OK.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: This works for me with `sleep(0)` too. And I don't see why it wouldn't -- all three tasks are runnable so `sleep(0)` will run each of them (presumably in creation order). c1 will return, c2 and c3 will hang in `await sem.acquire()`.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: This can be 0, see my comment on `_wake_up_first()`.

File: Lib/asyncio/locks.py
Comment (gvanrossum) [inline Lib/asyncio/locks.py:422]: Unlike Lock, in Semaphore you may call release() repeatedly. If there are several waiters, some of which have already been woken up (but not removed their Future yet), this should wake up the next one. This actually makes the code simpler (you may want to keep some of the comments, paraphrased):
```suggestion
        for fut in self._waiters:
            if not fut.done():
                fut.set_result(True)
                break
```
This removes the need for at least one sleep(0.01) in my testing, maybe more.

**Review by gvanrossum — COMMENTED** [review]
I think I've cracked the sleep(0.01) mystery.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: For some reason this timeout still needs to be 0.01. I'll look into it.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_locks.py]: You could use `r1, r2 = ...` instead of `result =`; we know we'll get exactly two results.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_locks.py]: This is because of how `wait_for()` is implemented -- using `print()` I verified that at this point the semaphore is unlocked and its value is 1, and the `acquire()` call takes a fraction of a millisecond. But `wait_for()` requires two iterations through the event loop -- the first to call `_release_waiter()`, then the second to wake up `wait_for()`. I'd specify a large timeout (e.g. 1.0) so hiccups in CI are less likely to cause test flakes -- we now know that the test passes, and if it were to fail it would just hang forever.

File: Lib/asyncio/locks.py
Comment (cykerway) [reply Lib/asyncio/locks.py:422]: The name of `_wake_up_first` implies it only wakes up the _first_ waiter. If we wake up multiple waiters at once, how do we know they will resume from `await` in FIFO order?

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: The time `0.01` second was taken from existing tests. Why would we change this one to `1.0` but not do the same thing at other places? I'm not sure if tests run in parallel and this would make tests run much longer.

File: Lib/asyncio/locks.py
Comment (cykerway) [reply Lib/asyncio/locks.py]: `Semaphore` docstring says the counter can never go below zero. So `locked` has been updated.

File: Lib/test/test_asyncio/test_locks.py
Comment (cykerway) [reply Lib/test/test_asyncio/test_locks.py]: This is currently pending on the comment under `_wake_up_first`.

**Review by gvanrossum — CHANGES_REQUESTED** [review]
There are several more sleep(0.01) cases that should be sleep(0), and I had a suggestion for the wait_for(..., timeout=0.01) too.

File: Lib/asyncio/locks.py
Comment (cykerway) [reply Lib/asyncio/locks.py:422]: After some trials I think what actually happens is worse than my previous comment. This change would result in negative counter in certain cases and this is a bug. An example is when you use `0.01` in test file Line 850. To explain it, the two `release()` would each awake a waiter. Then the first waiter awakes a third waiter because the second waiter hasn't yet decreased the counter. But there really isn't a value for the third waiter.

The test can pass when using time `0` simply because the execution hasn't reached far enough into the bug. In other words, the test does not completely test the entire execution. This probably suggests using `sleep(0.01)` to test is better in some circumstance. Otherwise we'd have to use multiple `sleep(0)`.

**Comment by cykerway** [comment]
The suggested change on `_wake_up_first` seems to be buggy and it blocks other changes. See my comment at Line 411.

**Comment by gvanrossum** [comment]
> The suggested change on `_wake_up_first` seems to be buggy and it blocks other changes. See my comment at Line 411.

I don’t see that comment. Did you send it?

**Comment by cykerway** [comment]
I think I did send it. But I can paste it here.

> After some trials I think what actually happens is worse than my previous comment. This change would result in negative counter in certain cases and this is a bug. An example is when you use 0.01 in test file Line 850. To explain it, the two release() would each awake a waiter. Then the first waiter awakes a third waiter because the second waiter hasn't yet decreased the counter. But there really isn't a value for the third waiter.

> The test can pass when using time 0 simply because the execution hasn't reached far enough into the bug. In other words, the test does not completely test the entire execution. This probably suggests using sleep(0.01) to test is better in some circumstance. Otherwise we'd have to use multiple sleep(0).

**Comment by gvanrossum** [comment]
> I think I did send it. But I can paste it here.

GitHub must have eaten it (or you accidentally entered review mode and it's still pending).

I'll have to think about what you wrote, you may be right. But I sure don't like having to use sleep(0.01), it makes for flaky tests in CI (where the VMs regularly have clock hiccups -- test_asyncio is already one of the flakiest).

**Comment by cykerway** [comment]
> (or you accidentally entered review mode and it's still pending)

Yeah I double checked and I was silly. They should have been sent now.

**Comment by gvanrossum** [comment]
You are right that only one waiter should be woken up. I wrote a little story (https://gist.github.com/gvanrossum/b4029df464b6bc2b7b49da60411a7c3e) and have convinced myself that my suggestion was wrong. Tomorrow I hope to have time to look at the remaining 0.01 sleeps -- I want all of them to go away, possibly being replaced with multiple 0 sleeps. Even ones that were already there.

File: Lib/asyncio/locks.py
Comment (gvanrossum) [inline Lib/asyncio/locks.py:349]: Or you could just set it to an empty deque like before, and simplify the code below -- at the cost of making the code less like `Lock`, and more expensive to create (but that only matters if semaphores routinely die without ever having been waited for).

File: Lib/asyncio/locks.py
Comment (gvanrossum) [inline Lib/asyncio/locks.py]: Please restore this blank line. (A blank line following the first line of the docstring is mandatory, see PEP 257.) You might want to fix the release() docstring too, which violated this. (Yes, I know this PEP is applied inconsistently.)

**Comment by gvanrossum** [comment]
I'm sorry, I have too many other things to do right now. I'll get back to this soon enough. I hope you have a few days' patience.

**Comment by cykerway** [comment]
> I'm sorry, I have too many other things to do right now. I'll get back to this soon enough. I hope you have a few days' patience.

No worry, I have other things too. Communication is pretty much async. On the other hand, if all the remaining issue is `sleep(0)` vs `sleep(0.01)`, then those may worth another PR, to keep things simple and manageable.

**Comment by gvanrossum** [comment]
Here are a couple of loose thoughts.
- I am still stunned by how hard this is to get right. There is only a single `await` in the whole class! Andrew and Kumar together didn't get it right in the previous iteration, it was broken before then, you and I are still struggling.
- Maybe we should allow `self._value` to become negative, where `-N` means there are `N` waiters in the queue. That seems like a nice invariant. Maybe one reason we're struggling is that the current code doesn't represent the right invariant in its state.
- Comparing `Lock` and `Semaphore`, why does `acquire()` call `self._wake_up_first()` in the end (after decrementing `self._value`)? Maybe if we remove that, we can use the loop I proposed in `_wake_up_first()`.
- Oh, I know the answer. We are trying to enforce strict FIFO order. But if there are two waiting guests, A (first in line) and B (second in line), and two dining guests leave simultaneously, it shouldn't matter whether we seat A or B first -- there's a table for each, so we should be able to buzz both, as neither of them will have to wait once they come back to the maitre d'. That's what restaurants do. Sure, it means B may get seated before A, but that shouldn't be a problem -- A will still get seated right away when they show up.
- The event loop, assuming it runs runnable task in order, will run A before B anyway. So the only thing we get for our efforts is that it takes an extra event loop iteration before B gets made runnable -- hence the need to sleep 0.01. (Because on any given iteration, the event loop runs everything that's runnable at the start of the iteration -- tasks that become runnable as a result of running another task will run at the _next_ iteration. IIRC this is important to avoid certain starvation patterns.)
- If all is well, `Lock` can be replaced by a subclass of `Semaphore` that takes no `value` argument (but not in the same PR, since this shouldn't be backported, and there may be other pragmatic reasons against doing this).
- Regardless of whether we go with my approach or yours, there's a difference between `Lock` and `Semaphore` that is directly connected to the changes in implementation (in your version, the extra `_wake_up_first()` at the end of `acquire()`, in my version, the loop in `_wake_up_first()`).
- What should `locked()` return when there are open tables but all of them are already committed to a guest that's been buzzed? It currently returns False (there's an open table) but you cannot be seated (that table is already spoken for).

**Review by gvanrossum — COMMENTED** [review]
Two style nits.

**Comment by kumaraditya303** [comment]
> I am still stunned by how hard this is to get right. There is only a single await in the whole class! Andrew and Kumar together didn't get it right in the previous iteration, it was broken before then, you and I are still struggling.

Maybe you are saying that because GitHub shows that I am a co-author on the commit? It was added mistakenly by Andrew (something went wrong in git?). I did not participate in the PR nor the issue until you pinged me. I didn't participate in https://github.com/python/cpython/pull/31910.

Anyways I provided my own implementation here https://github.com/python/cpython/issues/90155#issuecomment-1252235717. It has comments on what happens on loop iteration which would make it easier for you understand.

**Comment by kumaraditya303** [comment]
Also I would prefer to have `asyncio.sleep(0)`s rather than a single `asyncio.sleep(<small guessed delay>)` as it can be flaky.  `asyncio.sleep(0)` makes reasoning of the code much easier IMO. @gvanrossum @cykerway  Let me know if you want me to take over from here and fixup the tests.

**Comment by gvanrossum** [comment]
> Maybe you are saying that because GitHub shows that I am a co-author on the commit? It was added mistakenly by Andrew (something went wrong in git?). I did not participate in the PR nor the issue until you pinged me. I didn't participate in #31910.

Oops, sorry, you're right. I'll edit the comment.

> Anyways I provided my own implementation here [#90155 (comment)](https://github.com/python/cpython/issues/90155#issuecomment-1252235717). It has comments on what happens on loop iteration which would make it easier for you understand.

Thank you. I'll review that after my first morning meeting.

File: Lib/asyncio/locks.py
Comment (cykerway) [reply Lib/asyncio/locks.py:349]: > at the cost of making the code less like `Lock`

That's exactly what I'm trying to avoid. Are you sure you not want to update this for `Lock` and `Semaphore` together, in a separate commit?

**Comment by cykerway** [comment]
> * Maybe we should allow `self._value` to become negative

I haven't seen a similar implementation before. I doubt if it could be an invention.

> * Comparing `Lock` and `Semaphore`, why does `acquire()` call `self._wake_up_first()` in the end (after decrementing `self._value`)?

`Lock` should call it too, but it's optimized out in `Lock` because no two tasks can hold the lock at the same time.

> * two dining guests leave simultaneously, it shouldn't matter whether we seat A or B first

Seating A before B looks safer to me. Trust me or not, people would expect the FIFO order to be the order of waiters in queue (if you read the first post in Issue 90155).

> * If all is well, `Lock` can be replaced by a subclass of `Semaphore` that takes no `value` argument

`Lock` is basically `Semaphore(1)`, yet some implementation only allows unlock from the same task that locks.

> * What should `locked()` return when there are open tables but all of them are already committed to a guest that's been buzzed?

This is a semantic issue. Other semaphores I read provide methods to get the value and the number of waiters. We can provide them too if we want. But I prefer doing it in another commit. This one should be kept as small as possible since we don't even know if it works or not. We'd have a hard time reverting a big chunk of code if we make extensive change and its most basic functionality is proven flawed.

**Comment by cykerway** [comment]
> Also I would prefer to have `asyncio.sleep(0)`s rather than a single `asyncio.sleep(<small guessed delay>)` as it can be flaky. `asyncio.sleep(0)` makes reasoning of the code much easier IMO. @gvanrossum @cykerway Let me know if you want me to take over from here and fixup the tests.

`asyncio.sleep(0)` looks good to me, since testing each iteration is the most exact when you know what's happening. But this `0` vs `0.01` problem is not rooted in `Semaphore`. There are tens of `sleep(0.01)` in several files if you grep the tests. Why don't we get something running first then spend time optimizing it?

File: Lib/asyncio/locks.py
Comment (gvanrossum) [reply Lib/asyncio/locks.py:349]: Well, the original Lock code initialized the waiters queue lazily, and this makes sense to me -- Locks are supposed to be lightweight, and if you're just a little bit lucky a Lock may never be contended, hence saving the allocation and deallocation of an empty deque. OTOH the original Semaphore code always created a deque, and that makes sense to me too -- it's not as heavily used as Lock, so optimization vs. clean code is weighed differently, and I expect that the use cases for Semaphores (other than Semaphore(1) as a Lock) tend to be such that they will need the deque anyway.

But in the end I'm fine with your choice here.

File: Lib/asyncio/locks.py
Comment (gvanrossum) [reply Lib/asyncio/locks.py:422]: Yeah, I think I see the problem with my suggestion now. We *either* wake up the next waiter that isn't done yet in `release()`, and then `acquire()` only needs to wake someone when it is cancelled; *or* we do it your way. Both are correct. The difference is that your way requires two event loop iterations to wake two waiters, even in the test case that does `release(); release()`. The other way requires only one event loop iteration but relies on the event loop running tasks in order. So let's do it your way, but use multiple `sleep(0)` calls to make the tests pass instead of `sleep(0.01)` -- at least in the new tests that you've added, and in those tests that had a single `sleep(0)` before. Existing tests that were using `sleep(0.01)` can be fixed in a separate PR.

File: Lib/test/test_asyncio/test_locks.py
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_locks.py]: Because this test doesn't sleep -- if the test passes the timeout isn't reached. My explanation was not very clear -- the problem is not that `wait_for()` requires two event loop iterations (though it does), but that the timeout means something completely different than the sleep time in other places.

**Review by gvanrossum — COMMENTED** [review]
I'd still like to review whether now I understand your original fix -- if we want to backport this it's more convincing if we don't rewrite so much of the code.

**Review by cykerway — COMMENTED** [review]
Replaced `sleep(0.01)` with `sleep(0)` at multiple sites. Also changed `wait_for` timeout to `1.0`.

**Review by gvanrossum — APPROVED** [review]
Thanks, I think that's a wrap. I'll land this tomorrow, and mark it as a backport for 3.11, 3.10 and 3.9. Though I expect that 3.11 will be release without the fix -- I don't want to bother the release manager this close to the release, the fix will go into 3.11.1.

**Comment by gvanrossum** [comment]
I owe you some responses.

> > * Maybe we should allow `self._value` to become negative
> 
> I haven't seen a similar implementation before. I doubt if it could be an invention.

Probably not a good idea. :-) I do think I'd like to see a `.value()` or perhaps `.level()` method, but that's for a separate PR, and 3.12 only (it'd be a new feature).

> > * Comparing `Lock` and `Semaphore`, why does `acquire()` call `self._wake_up_first()` in the end (after decrementing `self._value`)?
> 
> `Lock` should call it too, but it's optimized out in `Lock` because no two tasks can hold the lock at the same time.

So for `Lock` it will never do anything.

> > * two dining guests leave simultaneously, it shouldn't matter whether we seat A or B first
> 
> Seating A before B looks safer to me. Trust me or not, people would expect the FIFO order to be the order of waiters in queue (if you read the first post in Issue 90155).

These are slightly different cases though. I agree that *if anyone else is waiting* (and in that example there are 20 tasks waiting) all waiting guests should be seated first. But if A and B are the first two in line, and suddenly two guests leave, we can definitely seat both A and B, and it shouldn't matter who we seat first. Though if some tables are more desirable, A might object that B gets the better table.

A different way to look at it is that if we make both A and B "runnable", in that order, the scheduler will resume A before B anyway (see [this code](https://github.com/python/cpython/blob/b4f5f07d076d7e7d825306981108dbb7120d7377/Lib/asyncio/base_events.py#L1889-L1912) in the event loop). The current algorithm makes only A runnable, and once A runs it makes B runnable (L396-7 in your code).

Given the immense complexity of the code I will give this a rest, but I wasn't referring to the case from the top of the issue (where a task that releases and then immediately acquires again manages to skip the queue).

> > * If all is well, `Lock` can be replaced by a subclass of `Semaphore` that takes no `value` argument
> 
> `Lock` is basically `Semaphore(1)`, yet some implementation only allows unlock from the same task that locks.

But ours doesn't care.

> > * What should `locked()` return when there are open tables but all of them are already committed to a guest that's been buzzed?
> 
> This is a semantic issue. Other semaphores I read provide methods to get the value and the number of waiters. We can provide them too if we want. But I prefer doing it in another commit. This one should be kept as small as possible since we don't even know if it works or not. We'd have a hard time reverting a big chunk of code if we make extensive change and its most basic functionality is proven flawed.

Indeed, I have come to the same conclusion (he code is already pretty substantial).

FWIW in a pre-emptive threading system (like Python's threading module) there's an argument that there's not much point in providing a method to tell whether a lock is locked since the value could change a few nanoseconds later. However in asyncio this is not the case -- the state of the lock remains the same until the next `await` call, or until you next call `release()` on it. So there is a stronger case IMO that `locked()` should return `True` if there are any (uncancelled) waiters. That's not what it's traditionally returned, so changing it would be a new feature.

Great job on the code and especially on all the new tests with all those subtle test cases, and thanks for bearing with me while I re-educated myself about semaphore implementation!

**Comment by kumaraditya303** [comment]
FYI 3.9 is now is in security fixes only so most likely this cannot be backport to 3.9.

**Comment by gvanrossum** [comment]
Okay, we'll go with this version.

**Comment by gvanrossum** [comment]
@cykerway Do you want to do the 3.9 backport by hand? We can also leave it be, 3.9 is in security-fix mode anyways.

**Comment by cykerway** [comment]
> Do you want to do the 3.9 backport by hand? We can also leave it be, 3.9 is in security-fix mode anyways.

No idea about that. I'm not quite familiar with the project management. I don't even know what the backport conflict is. Perhaps @kumaraditya303 is better on that topic.

**Comment by kumaraditya303** [comment]
3.9 is in security fixes only now, so better just leave it as it is, no need to backport.

**Comment by gvanrossum** [comment]
Okay, so after merging this I _still_ couldn't stop thinking about it, and I came up with a scenario where this is substantially worse (orders of magnitude) than before.

The [test program](https://gist.github.com/gvanrossum/02d9e4ef3c4f277050bf18fee0a3f189) spawns oodles of trivial tasks and tries to rate-limit them by making each task acquire a semaphore first. The semaphore allows 50 tasks at a time. With python 3.11rc2, on my Mac it does nearly 25,000 iterations per second. With the new code it does about 900, or about 27 times slower.

The first 50 tasks take the easy path through acquire(), every following task gets put in the queue first. What makes it so slow is that once 50 tasks have acquired the semaphore in a single event loop iteration, they will also release it all in a single iteration -- but the new algorithm only wakes up the first waiting task, and the next 49 release() calls do nothing to the queue. So from then on we only wake up one task per iteration.

Of course, it's easy to complain when the broken code is fast. :-) But I think we can do better.

**Comment by cykerway** [comment]
Looks like the slowdown was caused by the context switch between tasks. I'm not familiar enough with the loop to estimate how much overhead there is when control is passed to loop then passed back. The new version implements fifo order without depending on loop guarantees and this is safer. It sacrifices performance because it is not aided by the loop.

There are several things here: correctness, performance, fifo. We definitely want correctness, and there is a tradeoff between performance and fifo. If you try my first version it runs almost as fast as the old one before this commit and it doesn't not seem to have starvation problem but it's not obviously fifo. To tackle this tradeoff there should be some explicit guarantee from the loop yet I haven't seen it.

**Comment by gvanrossum** [comment]
From reading the code I know that the event loop absolutely calls callbacks that were scheduled using `call_soon()` in the order in which they were registered. I know that uvloop also uses this order. I feel that it's silly not to rely on this, since it gives us an important tool to fix the slowdown here. We should probably add a comment that states the dependency on rhis loop property.

**Comment by kumaraditya303** [comment]
Can you create a new issue to discuss this? Your gist program looks like the worst case scenario for this as it does no io and just yields. Also what is the throughput impact when you do some real work in function?

**Comment by gvanrossum** [comment]
Actually the behavior *is* guaranteed. Under [Scheduling Callbacks](https://docs.python.org/3.12/library/asyncio-eventloop.html#scheduling-callbacks) I read

> Callbacks are called in the order in which they are registered.

So we can definitely rely on this. (Sorry I hadn't mentioned this earlier, I wasn't actually aware of the guarantee, just of how the code works.)

Something that doesn't affect correctness but may affect performance is that the event loop goes through a cycle:

- wait for I/O events (*)
- register callbacks for I/O events
- call all callbacks that are registered and ready at this point (I/O and otherwise)
- go back to top

(*) The timeout for the I/O wait is zero if there are callbacks that are immediately ready, otherwise the time until the first callback scheduled for a particular time in the future (`call_later()`).

The I/O wait has expensive fixed overhead, so we want to call as many callbacks in a single iteration as possible.

Therefore I think a it behooves us to make all futures ready for which we have place. I think it can be like this abstract algorithm:

- Definitions:
  - Level: L = self._value
  - Waiters: W = self._waiters
  - Ready: R = [w for w in W if w.done() and not w.cancelled()]
  - Cancelled: C = [w for w in W if w.cancelled()]
  - Blocked: B = [w for w in W if not w.done()]
  - Note that R, C and B are views on W, not separate data structures
- Invariant that should hold at all times:
  - L >= |R|
    (I.e., we should not promise more guests to seat than we have open tables)
- Operations:
  - Equalize: while |B| > 0 and L >= |R|: make the first item of B ready (move it to R)
  - Release: L++; Equalize
  - Acquire:
    - if L > 0 and |R| == 0 and |B| == 0: L--; return
    - create a future F, append it to B, await it
    - when awaken (with or without exception):
      - assertion: F should be in either R or C
      - remove F from W (hence from R or C)
      - if no exception caught: L--; Equalize; return
      - if CancelledException caught and F.cancelled(): return
      - if CancelledException caught and not F.cancelled(): Equalize; return
      - (other exceptions are not expected and will bubble out unhandled)

**Comment by gvanrossum** [comment]
(It's not really resolved, but I needed to give a reason and that was less wrong than the other options GitHub gave me.)

**Pull request merged**
